### Solution Set for Shopify Intern
#### Name: Zifan Pan
#### Date: May 9, 2021

#### Question 1

In [1]:
import numpy as np
import pandas as pd

In [5]:
df = pd.read_csv('question_1_data.csv')

In [11]:
#check the format of data
df.head(10)

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0         1       53      746           224            2           cash   
1         2       92      925            90            1           cash   
2         3       44      861           144            1           cash   
3         4       18      935           156            1    credit_card   
4         5       18      883           156            1    credit_card   
5         6       58      882           138            1    credit_card   
6         7       87      915           149            1           cash   
7         8       22      761           292            2           cash   
8         9       64      914           266            2          debit   
9        10       52      788           146            1    credit_card   

            created_at  
0  2017-03-13 12:36:56  
1  2017-03-03 17:38:52  
2   2017-03-14 4:23:56  
3  2017-03-26 12:43:37  
4   2017-03-01 4:35:11  
5  2017-03-14 15:25:01  
6  2017-03-01 21:37:57  
7   2017-03-08 2:05:38  
8  2017-03-17 20:56:50  
9  2017-03-30 21:08:26

In [10]:
#find the original average order value
df['order_amount'].mean()

3145.128

The average order value is 3145.128, which is relatively higher than expected. From the first ten data above, we notice that the order amount for each store is significantly less than the average value of 3145.128. Thus there must be some abnormal outliers which are higher than 3145. 

In [15]:
#Filter out data that are above 3145
df[df['order_amount'] > 3145]

order_id  shop_id  user_id  order_amount  total_items payment_method  \
15          16       42      607        704000         2000    credit_card   
60          61       42      607        704000         2000    credit_card   
160        161       78      990         25725            1    credit_card   
490        491       78      936         51450            2          debit   
493        494       78      983         51450            2           cash   
...        ...      ...      ...           ...          ...            ...   
4646      4647       42      607        704000         2000    credit_card   
4715      4716       78      818         77175            3          debit   
4868      4869       42      607        704000         2000    credit_card   
4882      4883       42      607        704000         2000    credit_card   
4918      4919       78      823         25725            1           cash   

               created_at  
15     2017-03-07 4:00:00  
60     2017-03-04 4:00:00  
160    2017-03-12 5:56:57  
490   2017-03-26 17:08:19  
493   2017-03-16 21:39:35  
...                   ...  
4646   2017-03-02 4:00:00  
4715   2017-03-05 5:10:44  
4868   2017-03-22 4:00:00  
4882   2017-03-25 4:00:00  
4918  2017-03-15 13:26:46  

[63 rows x 7 columns]

We notice that shop #42 selled 2000 items for a total amount of 704000, which is not normal compared to other stores. Besides, shop #78 selled one pair of sneakers for 25725, which is not a common situation too. The data from these stores should not be included if we are calculating the average order value under common situations.  

In [19]:
#find which stores provided the abnormal data point,
# and count how many times the abnormal data appeared
df[df['order_amount'] > 3145].groupby('shop_id').count()

order_id  user_id  order_amount  total_items  payment_method  \
shop_id                                                                 
42             17       17            17           17              17   
78             46       46            46           46              46   

         created_at  
shop_id              
42               17  
78               46

Based on the above table obtained, we notice that there are only two shops which gives the outliers: shop #42 and shop #78, and they showed 17 and 46 times respectively. 

In [21]:
#remove these data points
df2 = df[df['order_amount'] < 3145]

In [22]:
df2

order_id  shop_id  user_id  order_amount  total_items payment_method  \
0            1       53      746           224            2           cash   
1            2       92      925            90            1           cash   
2            3       44      861           144            1           cash   
3            4       18      935           156            1    credit_card   
4            5       18      883           156            1    credit_card   
...        ...      ...      ...           ...          ...            ...   
4995      4996       73      993           330            2          debit   
4996      4997       48      789           234            2           cash   
4997      4998       56      867           351            3           cash   
4998      4999       60      825           354            2    credit_card   
4999      5000       44      734           288            2          debit   

               created_at  
0     2017-03-13 12:36:56  
1     2017-03-03 17:38:52  
2      2017-03-14 4:23:56  
3     2017-03-26 12:43:37  
4      2017-03-01 4:35:11  
...                   ...  
4995  2017-03-30 13:47:17  
4996  2017-03-16 20:36:16  
4997   2017-03-19 5:42:42  
4998  2017-03-16 14:51:18  
4999  2017-03-18 15:48:18  

[4937 rows x 7 columns]

In [25]:
#check if there are any missing points
df2.isnull().sum()

order_id          0
shop_id           0
user_id           0
order_amount      0
total_items       0
payment_method    0
created_at        0
dtype: int64

In reality, the price for a pair of normal sneakers would be around 90-150 dollars depends on the specific model. Therefore, it would be reasonable if the metric is dollar (USD/CAD/etc.). 

Now find the average order value (AOV):

In [46]:
#find AOV
df2['order_amount'].mean() 

302.58051448247926

In [34]:
df2['order_amount'].count() #total number of transactions

4937

In [48]:
df2['total_items'].sum() #average number of sneakers sold per store

9848

In [65]:
df2['shop_id'].nunique() #total number of stores left

99

In [74]:
df2['total_items'].mean() #average number of sneakers sold per transaction

1.9947336439133077

In [35]:
df2['order_amount'].describe() #statistics of AOV

count    4937.000000
mean      302.580514
std       160.804912
min        90.000000
25%       163.000000
50%       284.000000
75%       387.000000
max      1760.000000
Name: order_amount, dtype: float64

In [85]:
df2['user_id'].nunique() #how many individuals has made at least one purchase

300

In [88]:
df2['user_id'].value_counts().max() #maximum number of transaction times per user 

28

In [89]:
df2['user_id'].value_counts().min() #minimum number of transaction times per user 

7

In [103]:
df[df['order_amount'] > 3145].drop(['order_id','user_id'], axis=1).groupby('shop_id').sum() #large transaction amounts

order_amount  total_items
shop_id                           
42           11968000        34000
78            2263800           88

In short, the average order value across all 4937 transactions was 302.58 dollars, meaning on average Shopify can expected the shoe stores to sell 2 sneakers and have 302.58 dollars of income per transaction. During this 30-day window, all 99 stores sold a total number of 9848 sneakers, where the lowest transaction amount was 90 dollars, and the highest transaction amount was 1760 dollars. The median of all transactions was 284 dollars, but the mean was 302.58 dollars, so that the data is right-skewed, which means the stores should expect most of consumers spent less than 302.58 dollars each time they purchase snearkers. Besides, not all consumers has made a purchase, and some consumers purchased multiple times. Some user(s) made at most 28 times of transactions and some user(s) made at least 7 times of transactions. Lastly, regrading the large transactions, under the premise that the data is true, it would be benefical if store #42 and store #78 can seek the long-term cooperations with them. 